service/history/mutableStateBuilder.go

```diff
@@ -279,6 +279,9 @@ func (e *mutableStateBuilder) Load(
 	e.pendingActivityInfoIDs = state.ActivityInfos
 	for _, activityInfo := range state.ActivityInfos {
 		e.pendingActivityIDToEventID[activityInfo.ActivityId] = activityInfo.ScheduleId
+		if (activityInfo.TimerTaskStatus & timerTaskStatusCreatedHeartbeat) > 0 {
+			e.pendingActivityTimerHeartbeats[activityInfo.ScheduleId] = getLastHeartbeatTimeStamp(activityInfo)
```

Comment:
This change mostly makes sense except this part where we are setting the VisibilityTS to be lastHeartbeatTS which seems incorrect.  The whole point of upstream caller is to detect duplicate timer task creation and this change might cause problems with that.
Why can't we persist visibilityTS for heartbeat when a new heartbeat timer is created?  New heartbeat timer creation means we are already doing a CAS write.  We should include a dedupe identifier (visibilityTS, txID, etc) within mutable state instead on relying on this in-memory state which is very error prone.
But that might be a bigger change.  Let's talk more tomorrow.

Reply:
addressed feedback and am assigning time.Time{} instead, which should be safe and avoid unnecessary drops